tests/compose-spec_test.go

```diff
@@ -0,0 +1,227 @@
+        "services": {
+            "a": {
+              "image": "alpine"
+            }
+        },
+        %s
+        }`, spec)
+}
+
+func createComposeSpecService(serviceSpec string) string {
+	return fmt.Sprintf(`{
+    "services": {
+      "service-foo": {
+          %s
+      }
+    }
+  }`, serviceSpec)
+}
+
+func TestCompileComposeSpec(t *testing.T) {
+	_, err := jsonschema.Compile("../schema/compose-spec.json")
+	assert.NilError(t, err)
+}
+
+func TestComposeSpec(t *testing.T) {
+	tests := []struct {
+		name        string
+		spec        string
+		assertError bool
+	}{
+		{
+			name: "network ipv4 and ipv6 subnet",
+			spec: createComposeSpec(`
+        "networks": {
+            "front-tier": {
+                "ipam": {
+                    "driver": "default",
+                    "config": [
+                        {"subnet": "172.16.238.0/24"},
+                        {"subnet": "0.0.0.0/64"},
+                        {"subnet": "2001:3984:3989::/64"},
+                        {"subnet": "172.16.238.0/24"},
+                        {"subnet": "2001:3984:3989::/64"},
+                        {"subnet": "2001:3984:3989::/64"},
+                        {"subnet": "1080:0:0:0:8:800:200C:417A/64"},
+                        {"subnet": "FF01:0:0:0:0:0:0:101/64"},
+                        {"subnet": "0:0:0:0:0:0:0:1/64"},
+                        {"subnet": "0:0:0:0:0:0:0:0/64"},
+                        {"subnet": "::/64"},
+                        {"subnet": "1080::8:800:200C:417A/64"},
+                        {"subnet": "::1/64"}
+                    ]        }
+            }
+        }`),
+		},
+		{
+			name: "expose port has right format",
+			spec: createComposeSpecService(`
+        "expose": [
+            "12345-12345/tcp",
+            "12345-12345/udp",
+            "12345/udp",
+            "1234/udp",
+            "1/udp",
+            "1-1/udp"
+        ]`),
+		},
+		{
+			name: "expose port is to long",
+			spec: createComposeSpecService(`
+        "expose": [
+            "123456"
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "expose second port is too long",
+			spec: createComposeSpecService(`
+        "expose": [
+            "1234-123456"
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "expose missing first port",
+			spec: createComposeSpecService(`
+        "expose": [
+            "-123"
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "expose with wrong protocol",
+			spec: createComposeSpecService(`
+        "expose": [
+            "123-123/random"
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "ports must have the [HOST:]CONTAINER[/PROTOCOL] format",
+			spec: createComposeSpecService(`
+        "ports": [
+            "3000",
+            "3000-3005",
+            "8000:8000",
+            "9090-9091:8080-8081",
+            "49100:22",
+            "8000-9000:80",
+            "127.0.0.1:8001:8001",
+            "127.0.0.1:5000-5010:5000-5010",
```

Comment:
Probably need a few IPv6 test cases here, which I think might fail the current regex

(but I assume engine supports)